keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,966 @@
+    B[Candidate] --> |Elected| C[Leader]
+    C --> |Renew Leader Lease| C
+    C -->|End of Term / Leader Lease Expired| D[Shutdown]
+    D[Shutdown] -.-> |Restart| A
+```
+
+### Enabling on a component
+
+Components with a `--leader-elect-resource-lock` flag (kube-controller-manager,
+ kube-scheduler) will accept `coordinatedleases` as a resource lock type.
+
+### Comparison of leader election
+
+|                 | Lease Based Leader Election      | Coordinated Leader Election                                                    |
+| --------------- | -------------------------------- | ------------------------------------------------------------------------------ |
+| Lock Type       | Lease                            | Lease                                                                          |
+| Claimed by      | Component instance               | Election Coordinator. (Lease is claimed for to the elected component instance) |
+| Renewed by      | Component instance               | Component instance                                                             |
+| Leader Criteria | First component to claim lease   | Best leader from available candidates at time of election                      |
+| Preemptable     | No                               | Yes, Collaboratively. (Coordinator marks lease as "end of term". Component instance voluntarily stops renewing) |
+
+### User Stories (Optional)
+
+#### Story 1
+
+A cluster administrator upgrades a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first and second nodes are upgraded, any components that were leaders
+  will typically loose the lease during the node downtime
+  - If one happens to retain it's lease, it will be preempted by the coordinated
+    election controller after it updates it's identity lease with new version
+    information
+- When the third node is upgraded, all components will be at the new version and one
+  will be elected
+
+#### Story 2
+
+A cluster administrator rolls back a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first node is rolled back, any components that were leaders
+  will typically loose the lease during the node downtime
+- Once one of the components updates its identity lease with new version
+  information, the coordinated election controller will preempt the current
+  leader so that this lower version component becomes leader.
+- When the remaining two nodes can rollback, the first node will typically
+  remain leader, but if a new election occurs, the available older version
+  components will be elected.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
Another risk that we should mention here is performance/cost related.

Leader-election suddenly becomes visibly more expensive, because so far there was effectively a single update of an object per <lease-duration-period>/2, with this change we have additional <number of instances> of objects, each of them is being updated.

I can definitely imagine things like leader-election PL suddenly having too few shares in certain clusters etc.

@linxiulei - FYI

Reply:
FWIW - for now nothing in the core is watching lease's, which will change with the proposal.

[I've historically seen the visible watch-delay on leases in very large clusters, I think we addressed that, but definitely this is the next risk I see...]

Reply:
If I understand this KEP correctly, non-leaders will also periodically renew their identity leases. In terms of request count, it will be 2x but UPDATEs tend to take more time to process so in practice it might be >3x. One possible optimization can be that the non-leaders only renew their identity leases when they notice the current leader lease should be preempted by them per their preemption strategy (it could be also simply checking if there are any differences in lease annotations.)

Reply:
> with this change we have additional of objects, each of them is being updated.

I agree this is a risk, I'll add this to the risk section of this KEP so we can discuss in more detail.

Note that node leases don't change, only leader leases for control plane components, so we're talking about only a few additional leases here.

> I can definitely imagine things like leader-election PL suddenly having too few shares in certain clusters etc.

I think we're proposing about the same lease renewal overhead as adding 3 more nodes to a cluster.

Reply:
> FWIW - for now nothing in the core is watching lease's, which will change with the proposal.

UVIP does this with apiserver identity leases, but it's in alpha, so we haven't observed the impact of it in production environments yet.  But we should bring this risk up on that KEP as well, because it is likely to surface this issue before this KEP does.

Reply:
> One possible optimization can be that the non-leaders only renew their identity leases when they notice the current leader lease should be preempted by them per their preemption strategy

Yeah, or we could adjust the lease durations and renewal frequency.  I see component identity as something that might have other uses in the future (this has already happened with apiserver identity, even though it's still in alpha)

Reply:
> only leader leases for control plane components, so we're talking about only a few additional leases here.

I feel that the solution should be extensible to all controllers , not limited to control plane controllers.

Reply:
> I feel that the solution should be extensible to all controllers , not limited to control plane controllers.

Ah, yes.  The coordinated election would be available to all controllers-- anyone writing a custom controller could use it.   What I meant to say here is that we wouldn't use it for node leases since node leases are used for liveness purposes not for election purposes.